File: pkg/templates/typescript/cua/providers/anthropic.ts
/**
 * Anthropic CUA provider adapter.
 *
 * Uses the Anthropic SDK's beta computer-use API with Claude models.
 */

import { Anthropic } from '@anthropic-ai/sdk';
import type { CuaProvider, TaskOptions, TaskResult } from './index';

function getSystemPrompt(): string {
  const date = new Date().toLocaleDateString('en-US', { weekday: 'long', month: 'long', day: 'numeric', year: 'numeric' });
  return `<SYSTEM_CAPABILITY>
* You are utilising an Ubuntu virtual machine with internet access.
* When you connect to the display, CHROMIUM IS ALREADY OPEN.
* If you need to navigate to a new page, use ctrl+l to focus the url bar and then enter the url.
* After each step, take a screenshot and carefully evaluate if you have achieved the right outcome.
* Explicitly show your thinking: "I have evaluated step X..." If not correct, try again.
* Only when you confirm a step was executed correctly should you move on to the next one.
* The current date is ${date}.
</SYSTEM_CAPABILITY>

<IMPORTANT>
* When using Chromium, if a startup wizard appears, IGNORE IT.
* Click on the search bar and enter the appropriate URL there.
</IMPORTANT>`;
}

type BetaMessageParam = Anthropic.Beta.Messages.BetaMessageParam;
type BetaContentBlockParam = Anthropic.Beta.Messages.BetaContentBlockParam;

export class AnthropicProvider implements CuaProvider {
  readonly name = 'anthropic';
  private apiKey: string;

  constructor() {
    this.apiKey = process.env.ANTHROPIC_API_KEY ?? '';
  }

  isConfigured(): boolean {
    return this.apiKey.length > 0;
  }

  async runTask(options: TaskOptions): Promise<TaskResult> {
    const { query, kernel, sessionId, viewportWidth = 1280, viewportHeight = 800 } = options;
    const client = new Anthropic({ apiKey: this.apiKey, maxRetries: 4 });
    const model = options.model || 'claude-sonnet-4-6';

    const messages: BetaMessageParam[] = [{ role: 'user', content: query }];

    while (true) {
      const response = await client.beta.messages.create({
        max_tokens: 4096,
        messages,
        model,
        system: [{ type: 'text', text: getSystemPrompt(), cache_control: { type: 'ephemeral' } }],
        tools: [{
          type: 'computer_20251124',
          name: 'computer',
          display_width_px: viewportWidth,
          display_height_px: viewportHeight,
          display_number: 1,
        }],
        betas: ['computer-use-2025-11-24', 'prompt-caching-2024-07-31'],
        thinking: { type: 'enabled', budget_tokens: 1024 },
      });

      // Build assistant content for the messages array
      const assistantContent: BetaContentBlockParam[] = response.content.map(block => {
        if (block.type === 'thinking') {
          return { type: 'thinking' as const, thinking: block.thinking, signature: block.signature };
        }
        if (block.type === 'text') {
          return { type: 'text' as const, text: block.text };
        }
        if (block.type === 'tool_use') {
          return { type: 'tool_use' as const, id: block.id, name: block.name, input: block.input };
        }
        return block as unknown as BetaContentBlockParam;
      });
      messages.push({ role: 'assistant', content: assistantContent });

      if (response.stop_reason === 'end_turn') {
        const text = response.content
          .filter((b): b is Anthropic.Beta.Messages.BetaTextBlock => b.type === 'text')
          .map(b => b.text)
          .join('');
        return { result: text, provider: this.name };
      }

      // Process tool calls
      const toolResults: BetaContentBlockParam[] = [];
      for (const block of response.content) {
        if (block.type !== 'tool_use') continue;

        const input = block.input as Record<string, unknown>;
        const action = input.action as string;

        try {
          const screenshot = await this.executeAction(kernel, sessionId, action, input);
          toolResults.push({
            type: 'tool_result' as unknown as 'text',
            tool_use_id: block.id,
            content: [{ type: 'image', source: { type: 'base64', media_type: 'image/png', data: screenshot } }],
          } as unknown as BetaContentBlockParam);
        } catch (error) {
          toolResults.push({
            type: 'tool_result' as unknown as 'text',
            tool_use_id: block.id,
            content: [{ type: 'text', text: `Error: ${error instanceof Error ? error.message : String(error)}` }],
            is_error: true,
          } as unknown as BetaContentBlockParam);
        }
      }

      if (toolResults.length > 0) {
        messages.push({ role: 'user', content: toolResults });
      } else {
        // No tool use and not end_turn — model is done
        const text = response.content
          .filter((b): b is Anthropic.Beta.Messages.BetaTextBlock => b.type === 'text')
          .map(b => b.text)
          .join('');
        return { result: text || '(no response)', provider: this.name };
      }
    }
  }

  private async executeAction(
    kernel: TaskOptions['kernel'],
    sessionId: string,
    action: string,
    input: Record<string, unknown>,
  ): Promise<string> {
    const computer = kernel.browsers.computer;

    switch (action) {
      case 'screenshot': break;
      case 'key': {
        const key = input.key as string;
        await computer.pressKey(sessionId, { keys: [this.mapKey(key)] });
        break;
      }
      case 'hold_key': {
        const key = input.key as string;
        const duration = (input.duration as number) ?? 500;
        await computer.pressKey(sessionId, { keys: [this.mapKey(key)], duration });
        break;
      }
      case 'type': {
        const text = input.text as string;
        await computer.typeText(sessionId, { text });
        break;
      }
      case 'cursor_position': break;
      case 'mouse_move': {
        const [x, y] = input.coordinate as [number, number];
        await computer.moveMouse(sessionId, { x, y });
        break;
      }
      case 'left_click':
      case 'right_click':
      case 'middle_click': {
        const [x, y] = input.coordinate as [number, number];
        const button = action === 'right_click' ? 'right' : action === 'middle_click' ? 'middle' : 'left';
        await computer.clickMouse(sessionId, { x, y, button });
        break;
      }
      case 'double_click': {
        const [x, y] = input.coordinate as [number, number];
        await computer.clickMouse(sessionId, { x, y, num_clicks: 2 });
        break;
      }
      case 'triple_click': {
        const [x, y] = input.coordinate as [number, number];
        await computer.clickMouse(sessionId, { x, y, num_clicks: 3 });
        break;
      }
      case 'left_click_drag': {
        const startCoordinate = input.start_coordinate as [number, number];
        const [ex, ey] = input.coordinate as [number, number];
        await computer.dragMouse(sessionId, {
          path: [
            [startCoordinate[0], startCoordinate[1]],
            [ex, ey],
          ],
        });
        break;
      }
      case 'scroll': {
        const [x, y] = input.coordinate as [number, number];
        const direction = input.direction as string;
        const amount = (input.amount as number) ?? 3;
        const deltaX = direction === 'left' ? -amount : direction === 'right' ? amount : 0;
        const deltaY = direction === 'up' ? -amount : direction === 'down' ? amount : 0;
        await computer.scroll(sessionId, { x, y, delta_x: deltaX, delta_y: deltaY });
        break;
      }
      case 'wait': {
        const duration = (input.duration as number) ?? 1000;
        await new Promise(r => setTimeout(r, duration));
        break;
      }
      default:
        throw new Error(`Unknown action: ${action}`);
    }

    // Take screenshot after every action
    await new Promise(r => setTimeout(r, 500));
    const resp = await computer.captureScreenshot(sessionId);
    const buf = Buffer.from(await resp.arrayBuffer());
    return buf.toString('base64');
  }

  private mapKey(key: string): string {
    const map: Record<string, string> = {
      Return: 'Return', Enter: 'Return', Backspace: 'BackSpace',
      Tab: 'Tab', Escape: 'Escape', space: 'space', Space: 'space',
      Up: 'Up', Down: 'Down', Left: 'Left', Right: 'Right',
      Home: 'Home', End: 'End', Page_Up: 'Prior', Page_Down: 'Next',
      ctrl: 'Control_L', Control_L: 'Control_L', alt: 'Alt_L', Alt_L: 'Alt_L',
      shift: 'Shift_L', Shift_L: 'Shift_L', super: 'Super_L', Super_L: 'Super_L',
    };
    // Handle combos like "ctrl+l"
    if (key.includes('+')) {
      return key.split('+').map(k => map[k.trim()] ?? k.trim()).join('+');
    }
    return map[key] ?? key;
  }
}

File: pkg/templates/python/cua/providers/anthropic.py
"""Anthropic CUA provider adapter using Claude computer-use API."""

from __future__ import annotations

import asyncio
import base64
import os
from datetime import datetime

import anthropic

from . import CuaProvider, TaskOptions, TaskResult

SYSTEM_PROMPT = """<SYSTEM_CAPABILITY>
* You are utilising an Ubuntu virtual machine with internet access.
* When you connect to the display, CHROMIUM IS ALREADY OPEN.
* If you need to navigate to a new page, use ctrl+l to focus the url bar and then enter the url.
* After each step, take a screenshot and carefully evaluate if you have achieved the right outcome.
* Only when you confirm a step was executed correctly should you move on to the next one.
* The current date is {date}.
</SYSTEM_CAPABILITY>

<IMPORTANT>
* When using Chromium, if a startup wizard appears, IGNORE IT.
* Click on the search bar and enter the appropriate URL there.
</IMPORTANT>"""

KEY_MAP = {
    "Return": "Return", "Enter": "Return", "Backspace": "BackSpace",
    "Tab": "Tab", "Escape": "Escape", "space": "space", "Space": "space",
    "Up": "Up", "Down": "Down", "Left": "Left", "Right": "Right",
    "Home": "Home", "End": "End", "Page_Up": "Prior", "Page_Down": "Next",
    "ctrl": "Control_L", "Control_L": "Control_L",
    "alt": "Alt_L", "Alt_L": "Alt_L",
    "shift": "Shift_L", "Shift_L": "Shift_L",
    "super": "Super_L", "Super_L": "Super_L",
}


def _map_key(key: str) -> str:
    if "+" in key:
        return "+".join(KEY_MAP.get(k.strip(), k.strip()) for k in key.split("+"))
    return KEY_MAP.get(key, key)


class AnthropicProvider:
    name = "anthropic"

    def __init__(self) -> None:
        self._api_key = os.environ.get("ANTHROPIC_API_KEY", "")

    def is_configured(self) -> bool:
        return len(self._api_key) > 0

    async def run_task(self, options: TaskOptions) -> TaskResult:
        client = anthropic.Anthropic(api_key=self._api_key, max_retries=4)
        model = "claude-sonnet-4-6"
        messages: list[dict] = [{"role": "user", "content": options.query}]

        date_str = datetime.now().strftime("%A, %B %d, %Y")
        system_prompt = SYSTEM_PROMPT.format(date=date_str)

        while True:
            response = await asyncio.to_thread(
                client.beta.messages.create,
                max_tokens=4096,
                messages=messages,
                model=model,
                system=[{"type": "text", "text": system_prompt, "cache_control": {"type": "ephemeral"}}],
                tools=[{
                    "type": "computer_20251124",
                    "name": "computer",
                    "display_width_px": options.viewport_width,
                    "display_height_px": options.viewport_height,
                    "display_number": 1,
                }],
                betas=["computer-use-2025-01-24", "prompt-caching-2024-07-31"],
                thinking={"type": "enabled", "budget_tokens": 1024},
            )

            assistant_content = []
            for block in response.content:
                if block.type == "thinking":
                    assistant_content.append({
                        "type": "thinking",
                        "thinking": block.thinking,
                        "signature": block.signature,
                    })
                elif block.type == "text":
                    assistant_content.append({"type": "text", "text": block.text})
                elif block.type == "tool_use":
                    assistant_content.append({
                        "type": "tool_use",
                        "id": block.id,
                        "name": block.name,
                        "input": block.input,
                    })

            messages.append({"role": "assistant", "content": assistant_content})

            if response.stop_reason == "end_turn":
                text = " ".join(
                    b.text for b in response.content if b.type == "text"
                )
                return TaskResult(result=text, provider=self.name)

            # Process tool calls
            tool_results = []
            for block in response.content:
                if block.type != "tool_use":
                    continue
                action = block.input.get("action", "")
                try:
                    screenshot = await self._execute_action(
                        options, action, block.input,
                    )
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": [{
                            "type": "image",
                            "source": {
                                "type": "base64",
                                "media_type": "image/png",
                                "data": screenshot,
                            },
                        }],
                    })
                except Exception as exc:
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": [{"type": "text", "text": f"Error: {exc}"}],
                        "is_error": True,
                    })

            if tool_results:
                messages.append({"role": "user", "content": tool_results})
            else:
                text = " ".join(
                    b.text for b in response.content if b.type == "text"
                )
                return TaskResult(result=text or "(no response)", provider=self.name)

    async def _execute_action(
        self, options: TaskOptions, action: str, params: dict
    ) -> str:
        computer = options.kernel.browsers.computer

        if action == "screenshot":
            pass
        elif action == "key":
            key = _map_key(params.get("key", ""))
            await asyncio.to_thread(
                computer.press_key, options.session_id, keys=[key]
            )
        elif action == "hold_key":
            key = _map_key(params.get("key", ""))
            duration = params.get("duration", 500)
            await asyncio.to_thread(
                computer.press_key, options.session_id,
                keys=[key], duration=duration,
            )
        elif action == "type":
            text = params.get("text", "")
            await asyncio.to_thread(
                computer.type_text, options.session_id, text=text,
            )
        elif action in ("left_click", "right_click", "middle_click"):
            x, y = params.get("coordinate", [0, 0])
            button = {"left_click": "left", "right_click": "right", "middle_click": "middle"}[action]
            await asyncio.to_thread(
                computer.click_mouse, options.session_id, x=x, y=y, button=button,
            )
        elif action == "double_click":
            x, y = params.get("coordinate", [0, 0])
            await asyncio.to_thread(
                computer.click_mouse, options.session_id, x=x, y=y, num_clicks=2,
            )
        elif action == "triple_click":
            x, y = params.get("coordinate", [0, 0])
            await asyncio.to_thread(
                computer.click_mouse, options.session_id, x=x, y=y, num_clicks=3,
            )
        elif action == "mouse_move":
            x, y = params.get("coordinate", [0, 0])
            await asyncio.to_thread(
                computer.move_mouse, options.session_id, x=x, y=y,
            )
        elif action == "left_click_drag":
            sx, sy = params.get("start_coordinate", [0, 0])
            ex, ey = params.get("coordinate", [0, 0])
            await asyncio.to_thread(
                computer.drag_mouse, options.session_id,
                path=[[sx, sy], [ex, ey]],
            )
        elif action == "scroll":
            x, y = params.get("coordinate", [0, 0])
            direction = params.get("direction", "down")
            amount = params.get("amount", 3)
            dx = -amount if direction == "left" else amount if direction == "right" else 0
            dy = -amount if direction == "up" else amount if direction == "down" else 0
            await asyncio.to_thread(
                computer.scroll, options.session_id,
                x=x, y=y, delta_x=dx, delta_y=dy,
            )
        elif action == "wait":
            duration = params.get("duration", 1000)
            await asyncio.sleep(duration / 1000)
        elif action == "cursor_position":
            pass
        else:
            raise ValueError(f"Unknown action: {action}")

        # Screenshot after every action
        await asyncio.sleep(0.5)
        resp = await asyncio.to_thread(
            computer.capture_screenshot, options.session_id,
        )
        return base64.b64encode(resp.read()).decode()
